<a href="https://colab.research.google.com/github/switlanakostyuk-ctrl/Apollo/blob/main/%22hw_10_4_apply%2C_groupby%2C_pivot_table_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [6]:
import pandas as pd
df = pd.read_csv('supermarket_sales.csv')
path = '/content/drive/MyDrive/supermarket_sales.csv'

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [7]:
profit_by_branch = df.groupby("Branch")["gross income"].sum()
profit_by_branch

,gross income
Branch,
A,5057.1605
B,5057.0320
C,5265.1765


In [8]:
best_branch = profit_by_branch.idxmax()
best_branch

'C'

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [9]:
city_of_best_branch = df[df["Branch"] == best_branch]["City"].iloc[0]
city_of_best_branch

'Naypyitaw'

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [10]:
invoices_by_category = pd.pivot_table(
    df,
    values="Invoice ID",
    index="Branch",
    columns="Product line",
    aggfunc="count"
)

invoices_by_category

Product line,Electronic accessories,Fashion accessories,Food and beverages,Health and beauty,Home and lifestyle,Sports and travel
Branch,,,,,,
A,60,51,58,47,65,59
B,55,62,50,53,50,62
C,55,65,66,52,45,45


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [11]:
best_branch_electronics = invoices_by_category["Electronic accessories"].idxmax()
best_branch_electronics

'A'

4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


Гіпотеза 1. Вплив статі на категорії товарів.

Яка різниця між витратами на жіночі і чоловічі товари?

In [13]:
gender_preferences = df.pivot_table(index='Product line', columns='Gender', values='Total', aggfunc='mean')
print(gender_preferences)

Gender                      Female        Male
Product line                                  
Electronic accessories  322.643125  316.691965
Fashion accessories     317.056250  291.079207
Food and beverages      368.565750  273.499125
Health and beauty       290.015414  348.099460
Home and lifestyle      380.213639  294.136241
Sports and travel       324.712739  340.360327


Бізнес-рішення: чоловіки витрачають більше в категорії "Health and beauty", а жінки в "Electronic accessories", ми можемо змінити таргет реклами в соцмережах або переставити полиці в магазині так, щоб популярні товари були на виду у відповідної аудиторії.

Гіпотеза 2. Ефективність покупок залежно від дня тижня (Weekday Analysis)
У які дні тижня середній рейтинг або сума чека падають?

In [16]:
df['Date'] = pd.to_datetime(df['Date'])
df['Weekday'] = df['Date'].dt.day_name()

In [17]:
weekday_analysis = df.groupby('Weekday')[['Total', 'Rating']].mean().sort_values(by='Total', ascending=False)
print(weekday_analysis)

                Total    Rating
Weekday                        
Saturday   342.200058  6.901829
Sunday     334.269868  7.011278
Thursday   328.617739  6.889855
Tuesday    325.836997  7.003165
Friday     316.016838  7.076259
Wednesday  305.812133  6.805594
Monday     303.192624  7.153600


 Субота: Найвищий чек ($342$), але один з найнижчих рейтингів ($6.9$).Понеділок: Найнижчий чек ($303$), але найвищий рейтинг ($7.15$)

Дані показують зворотну кореляцію: коли грошей багато (вихідні), задоволеність клієнтів падає.

Рішення: У суботу та неділю потрібно виводити більше персоналу на зміну (касирів, консультантів), щоб зменшити черги та підвищити якість сервісу. У понеділок, коли клієнтів мало, персонал працює краще — можна подумати про скорочення зміни або проведення навчання саме в цей час.

Гіпотеза 3. Ефективність покупок у вечірній час (Happy Hours)
Чи змінюється середній рейтинг після 18:00? Вечір — це пік покупок, але чи залишаються клієнти задоволеними?

In [21]:
def get_hour(time_str):
    return int(time_str.split(':')[0])

df['Hour'] = df['Time'].apply(get_hour)

In [22]:
df['Is Evening'] = df['Hour'] >= 18

In [23]:
evening_efficiency = df.groupby('Is Evening')[['Rating', 'Total']].mean()
print(evening_efficiency)

              Rating       Total
Is Evening                      
False       6.984701  325.823879
True        6.941993  315.656157


Вдень: Рейтинг 6.98, середній чек $325.82

Ввечері: Рейтинг 6.94, середній чек $315.65

Бізнес-рішення: середній чек ввечері нижчий, і клієнти менш задоволені (рейтинг падає).

Чому так? Ймовірно, ввечері люди забігають за дрібницями після роботи, а великі черги псують настрій.

Рішення: Запровадити каси самообслуговування, щоб пришвидшити потік і підняти рейтинг.